org/pr/android-diff-size.ts

```diff
@@ -0,0 +1,30 @@
+import {warn, danger} from "danger";
+
+export default async () => {
+    // Skip for release PRs
+    const githubLabels = danger.github.issue.labels;
+    const modifiedFiles = danger.git.modified_files;
+
+    if (githubLabels.length != 0) {
+        const releases = githubLabels.some(label => label.name.includes("Releases"));
+        if (releases) {
+            return;
+        }
+    }
+    
+    // Calculate the changes to Test files because we don't want to limit those
+    const modifiedStrings = modifiedFiles.filter((path: string) => path.includes("/src/test"));
+    let changesToTests = 0;
+    for (let file of modifiedStrings) {        
+        const stringDiffs = await danger.git.diffForFile(file)
+        const addedLines = stringDiffs.added.split("\n").filter(x => x.startsWith("+")).length;
+        const removedLines = stringDiffs.removed.split("\n").filter(x => x.startsWith("-")).length;
+        changesToTests += addedLines + removedLines;
+    }
+
+    // Warn when there is a big PR
+    let codeChanges = danger.github.pr.additions + danger.github.pr.deletions - changesToTests;
+    if (codeChanges > 300) {
```

Comment:
I'm curious to what the rationale/benefit is to "compute the diff size for tests, then substract that from the `pr.additions+pr.deletions` total") vs directly computing the diff size for non-tests?

In other words, why not directly run your code from lines 18-23 on the non-test files?

```suggestion
    // Calculate the changes to non-test files because we don't want to limit those
    const nonTestFiles = modifiedFiles.filter((path: string) => !path.includes("/src/test"));
    let codeChanges = 0;
    for (let file of nonTestFiles) {        
        const stringDiffs = await danger.git.diffForFile(file)
        const addedLines = stringDiffs.added.split("\n").filter(x => x.startsWith("+")).length;
        const removedLines = stringDiffs.removed.split("\n").filter(x => x.startsWith("-")).length;
        codeChanges += addedLines + removedLines;
    }

    // Warn when there is a big PR
    if (codeChanges > 300) {
```

Reply:
Oh, this is just for speed. The values of total additions and deletions are returned by the API and, in many PRs, they are all we need.

Reply:
Ah ok, so you estimated that there will be more non-test files than test files so it was quicker, got it!

That being said, while looking at the Danger docs to dig a bit deeper into the structure of the `diffForFile` return value, I also found [this `linesOfCode` method](https://github.com/danger/danger-js/blob/main/source/dsl/GitDSL.ts#L191) in the GitDSL exposed to Danger… which actually seems to do everything we want here – and actually a bit better, since it [also considers created files and deleted files in the count](https://github.com/danger/danger-js/blob/main/source/platforms/git/gitJSONToGitDSL.ts#L160-L184), which apparently we forgot to include here as we only consider `modifiedFiles`. I think it could allow us to simplify the code quite a lot here!🎉

Reply:
If I'm not mistaken that means that you could replace all the lines 15-23 with just `changesToTests = danger.git.linesOfCode('/src/test/**')`. I'm not sure if there's a way to negate a glob pattern (to directly get `codeChanges = danger.git.linesOfCode('!/src/test/**')` or something, assuming that's supported), but even use the positive glob then substract to the total would be both more accurate and readable than what we initially tried here, right?

Reply:
Ooooh apparently negative glob _is_ supported, since the [DangerJS code seems to rely](https://github.com/danger/danger-js/blob/main/source/platforms/git/gitJSONToGitDSL.ts#L161) on the [micromatch library](https://github.com/micromatch/micromatch) and in the README of that lib they have the following example:

```
console.log(micromatch(['foo', 'bar', 'baz', 'qux'], ['*', '!b*'])) //=> ['foo', 'qux']
```

I say it's worth a try, since I'd allow us to make our whole code logic down to a single `linesOfCode('!/src/test/**')` if that indeed works!

Reply:
Yup, it was great if we could use that API. Unluckily this has been implemented in a newer version of Danger than the one currently integrated in Peril. The version of the API we have available actually implements `linesOfCode`, but it is an older version which doesn't accept any parameter for filters yet :-(

There's an [open issue related to updating Danger in Peril](https://github.com/danger/peril/issues/470), but it seems nobody is currently actively working on it. It could be a nice HackWeek project :-)

Reply:
Ah, good catch. Dammit 😒  Let's keep that manual then (though it could be nice to add a comment with a link to that issue so we consider updating it once it's solved and shipped).

---

What do you think of including `createdFiles` and `deletedFiles` in addition to `modifiedFiles` when counting lines in tests? (This would probably look like this below but haven't tested if that's valid syntax). I'm especially thinking about the case that most PRs that currently a high count because of tests (that we now want to ignore) could be from new features, and thus potentially introducing new test files alongside the new feature, so that might actually be one of the most common case of the highest line counts due to tests in PRs…

```
const allFiles = (modifiedFiles + createdFiles + deletedFiles).filter((path: string) => path.includes("/src/test"));
```
Just hoping that `diffForFile(aDeletedFile)` is valid (and returns all its lines in the `removed:` key) and wouldn't e.g. throw an error for file not found, but I think that would work.

Reply:
Btw, while looking briefly into that update, I managed to implement branch support for the configuration file. Danger already has it, but Peril doesn't. I'll open a PR soon as I think it will make testing changes easier.

Reply:
![image](https://user-images.githubusercontent.com/216089/129370732-c205a977-216e-403e-aa2c-b12b176fcc48.png)

Reply:
> What do you think of including createdFiles and deletedFiles in addition to modifiedFiles when counting lines in tests?

Ah, good point! Thanks for catching it. Added in `0b507cf`